File: tests/cuda/lua/test_signal_io.lua
-- Signal io CUDA binding tests
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
--   ~/.luarocks/bin/busted tests/cuda/lua/test_signal_io.lua

local ok_ins, ins = pcall(require, "_insight")
if not ok_ins then
  print("SKIP: _insight module not available")
  os.exit(0)
end

local ok_gpu = pcall(function()
  ins.init()
end)
if not ok_gpu then
  print("SKIP: GPU backend not available")
  os.exit(0)
end

ins.set_device(ins.GPUPlace(0))

describe("Signal IO CUDA Tests", function()
  local tmpfile

  after_each(function()
    if tmpfile then
      os.remove(tmpfile)
      tmpfile = nil
    end
  end)

  it("write_read_bin_roundtrip", function()
    local data = ins.from_table({ 1.0, 2.0, 3.0, 4.0, 5.0 }):to(ins.GPUPlace(0))
    tmpfile = os.tmpname() .. ".bin"
    ins.signal.write_bin(data, tmpfile)
    local result = ins.signal.read_bin(tmpfile, "float64")
    assert.is_not_nil(result)
    assert.are.equal(5, result.numel)
  end)

  it("write_read_bin_float32", function()
    local data = ins.from_table({ 10.0, 20.0, 30.0 }):to(ins.GPUPlace(0))
    tmpfile = os.tmpname() .. ".bin"
    ins.signal.write_bin(data, tmpfile)
    local result = ins.signal.read_bin(tmpfile, "float32")
    assert.is_not_nil(result)
    assert.are.equal(3, result.numel)
  end)

  it("write_read_bin_int16", function()
    local data = ins.from_table({ 100, 200, 300, 400 }, ins.int16):to(ins.GPUPlace(0))
    tmpfile = os.tmpname() .. ".bin"
    ins.signal.write_bin(data, tmpfile)
    local result = ins.signal.read_bin(tmpfile, "int16")
    assert.is_not_nil(result)
    assert.are.equal(4, result.numel)
  end)

  it("write_read_bin_large", function()
    local t = {}
    for i = 1, 1024 do
      t[i] = math.sin(i)
    end
    local data = ins.from_table(t):to(ins.GPUPlace(0))
    tmpfile = os.tmpname() .. ".bin"
    ins.signal.write_bin(data, tmpfile)
    local result = ins.signal.read_bin(tmpfile, "float64")
    assert.is_not_nil(result)
    assert.are.equal(1024, result.numel)
  end)

  it("pack_bin", function()
    local data = ins.from_table({ 1.0, 2.0, 3.0 }):to(ins.GPUPlace(0))
    local packed = ins.signal.pack_bin(data, "float64")
    assert.is_not_nil(packed)
    assert.is_true(packed.numel > 0)
  end)

  it("unpack_bin", function()
    local data = ins.from_table({ 1.0, 2.0, 3.0, 4.0 }):to(ins.GPUPlace(0))
    tmpfile = os.tmpname() .. ".bin"
    ins.signal.write_bin(data, tmpfile)
    local raw = ins.signal.read_bin(tmpfile, "uint8")
    local result = ins.signal.unpack_bin(raw, "float64")
    assert.is_not_nil(result)
    assert.is_true(result.numel > 0)
  end)

  it("write_sigmf", function()
    local data = ins.from_table({ 1.0, 2.0, 3.0, 4.0 }):to(ins.GPUPlace(0))
    tmpfile = os.tmpname() .. ".sigmf-data"
    ins.signal.write_sigmf(data, tmpfile, {})
    assert.is_not_nil(tmpfile)
  end)
end)

File: tests/cuda/julia/test_signal_io.jl
# Signal io CUDA binding tests
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_signal_io.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

# Try GPU
try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0
failed = 0

function check(name, cond)
    global passed, failed
    if cond
        passed += 1
    else
        failed += 1
        println("FAIL: $name")
    end
end

function approx(a, b; atol=1e-10)
    return abs(a - b) < atol
end

# ============================================================================
# Signal I/O
# ============================================================================
println("=== Signal IO ===")

# write_bin / read_bin roundtrip (GPU write, CPU read)
data = Insight.from_data([1.0, 2.0, 3.0, 4.0, 5.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
tmpfile = tempname() * ".bin"
Insight.write_bin(tmpfile, data_gpu)
result = Insight.read_bin(tmpfile, dtype=Float64)
check("write_read_roundtrip", Insight.numel(result) == 5)
rm(tmpfile, force=true)

# write_bin / read_bin float32
data = Insight.from_data(Float32[10.0, 20.0, 30.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
tmpfile = tempname() * ".bin"
Insight.write_bin(tmpfile, data_gpu)
result = Insight.read_bin(tmpfile, dtype=Float32)
check("write_read_float32", Insight.numel(result) == 3)
rm(tmpfile, force=true)

# write_bin / read_bin int16
data = Insight.from_data(Int16[100, 200, 300, 400])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
tmpfile = tempname() * ".bin"
Insight.write_bin(tmpfile, data_gpu)
result = Insight.read_bin(tmpfile, dtype=Int16)
check("write_read_int16", Insight.numel(result) == 4)
rm(tmpfile, force=true)

# write_bin / read_bin large
large_data = [sin(i) for i in 1:1024]
data = Insight.from_data(large_data)
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
tmpfile = tempname() * ".bin"
Insight.write_bin(tmpfile, data_gpu)
result = Insight.read_bin(tmpfile, dtype=Float64)
check("write_read_large", Insight.numel(result) == 1024)
rm(tmpfile, force=true)

# pack_bin
data = Insight.from_data([1.0, 2.0, 3.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
packed = Insight.pack_bin(data_gpu)
check("pack_bin", Insight.numel(packed) > 0)

# unpack_bin
data = Insight.from_data([1.0, 2.0, 3.0, 4.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
tmpfile = tempname() * ".bin"
Insight.write_bin(tmpfile, data_gpu)
raw = Insight.read_bin(tmpfile, dtype=UInt8)
unpacked = Insight.unpack_bin(raw, Float64)
check("unpack_bin", Insight.numel(unpacked) > 0)
rm(tmpfile, force=true)

# write_sigmf
data = Insight.from_data([1.0, 2.0, 3.0, 4.0])
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
tmpfile = tempname() * ".sigmf-data"
Insight.write_sigmf(tmpfile, data_gpu)
check("write_sigmf", isfile(tmpfile))
rm(tmpfile, force=true)

# read_bin roundtrip data integrity
orig = [1.5, 2.5, 3.5, 4.5, 5.5]
data = Insight.from_data(orig)
data_gpu = Insight.to_array(data, Insight.GPUPlace(0))
tmpfile = tempname() * ".bin"
Insight.write_bin(tmpfile, data_gpu)
result = Insight.read_bin(tmpfile, dtype=Float64)
back = Insight.to_data(result)
check("roundtrip_integrity", all(approx.(orig, back)))
rm(tmpfile, force=true)

# ============================================================================
# Results
# ============================================================================
println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0
    exit(1)
end
